File: backend/app/middleware/security.py
from fastapi import Request, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.middleware.gzip import GZipMiddleware
from fastapi.middleware.trustedhost import TrustedHostMiddleware
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.types import ASGIApp
import time
import logging
from typing import List, Optional, Dict, Any, Callable, Awaitable
from slowapi import Limiter, _rate_limit_exceeded_handler
from slowapi.util import get_remote_address
from slowapi.errors import RateLimitExceeded
from slowapi.middleware import SlowAPIMiddleware

logger = logging.getLogger(__name__)

class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """Middleware for adding security headers to responses"""
    
    def __init__(
        self,
        app: ASGIApp,
        headers: Optional[Dict[str, str]] = None,
    ) -> None:
        super().__init__(app)
        self.headers = headers or self._get_default_headers()
    
    @staticmethod
    def _get_default_headers() -> Dict[str, str]:
        """Return default security headers"""
        return {
            "X-Frame-Options": "DENY",
            "X-Content-Type-Options": "nosniff",
            "X-XSS-Protection": "1; mode=block",
            "Referrer-Policy": "strict-origin-when-cross-origin",
            "Content-Security-Policy": "default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self' data:;",
            "Permissions-Policy": "camera=(), microphone=(), geolocation=()",
            "Strict-Transport-Security": "max-age=31536000; includeSubDomains",
        }
    
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint):
        response = await call_next(request)
        
        # Add security headers to the response
        for header, value in self.headers.items():
            response.headers[header] = value
            
        return response

class RateLimiterMiddleware:
    """Rate limiting middleware using slowapi"""
    
    def __init__(self, app: ASGIApp, rate_limit: str = "100/minute"):
        self.app = app
        self.limiter = Limiter(key_func=get_remote_address, default_limits=[rate_limit])
        
    async def __call__(self, scope, receive, send):
        if scope["type"] == "http":
            request = Request(scope, receive=receive)
            
            # Skip rate limiting for certain paths
            if not self._should_rate_limit(request):
                return await self.app(scope, receive, send)
                
            # Check rate limit
            try:
                # This is a simplified version - in practice, you'd use the limiter
                # to check and increment the rate limit counter
                await self.limiter.check(request)
            except RateLimitExceeded:
                raise HTTPException(status_code=429, detail="Too many requests")
                
        return await self.app(scope, receive, send)
    
    def _should_rate_limit(self, request: Request) -> bool:
        """Determine if the request should be rate limited"""
        # Skip rate limiting for health checks and static files
        skip_paths = ["/health", "/metrics", "/static/"]
        return not any(request.url.path.startswith(path) for path in skip_paths)

def setup_security_middleware(app: ASGIApp, **kwargs):
    """Set up all security-related middleware"""
    
    # Get CORS origins from settings
    cors_origins = [origin.strip() for origin in kwargs.get("cors_origins", "").split(",") if origin.strip()]
    
    # Add CORS middleware
    app.add_middleware(
        CORSMiddleware,
        allow_origins=cors_origins,
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
        expose_headers=["Content-Disposition"],
    )
    
    # Add security headers middleware
    app.add_middleware(SecurityHeadersMiddleware)
    
    # Add rate limiting
    rate_limit = kwargs.get("rate_limit", "100/minute")
    app.add_middleware(RateLimiterMiddleware, rate_limit=rate_limit)
    
    # Add GZip compression
    app.add_middleware(GZipMiddleware, minimum_size=1000)
    
    # Add trusted hosts middleware
    allowed_hosts = kwargs.get("allowed_hosts", ["*"])
    app.add_middleware(TrustedHostMiddleware, allowed_hosts=allowed_hosts)
    
    logger.info("Security middleware initialized")
    return app

File: backend/app/middleware/security.ts
import { Request, Response, NextFunction } from 'express';
import { validationResult } from 'express-validator';
import rateLimit from 'express-rate-limit';
import helmet from 'helmet';
import { v4 as uuidv4 } from 'uuid';
import crypto from 'crypto';

// Security headers middleware
export const securityHeaders = (req: Request, res: Response, next: NextFunction) => {
  // Generate nonce for CSP
  const nonce = crypto.randomBytes(16).toString('base64');
  
  // Set security headers
  res.set({
    'Content-Security-Policy': `
      default-src 'self';
      script-src 'self' 'nonce-${nonce}' 'strict-dynamic';
      style-src 'self' 'unsafe-inline';
      img-src 'self' data: https:;
      font-src 'self';
      connect-src 'self' ${process.env.API_BASE_URL || ''};
      frame-ancestors 'none';
      base-uri 'self';
      form-action 'self';
    `.replace(/\s+/g, ' ').trim(),
    'X-Content-Type-Options': 'nosniff',
    'X-Frame-Options': 'DENY',
    'X-XSS-Protection': '1; mode=block',
    'Referrer-Policy': 'strict-origin-when-cross-origin',
    'Permissions-Policy': 'geolocation=(), microphone=(), camera=()',
    'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload',
    'X-Request-Id': uuidv4(),
  });

  // Store nonce in response locals for use in templates
  res.locals.cspNonce = nonce;
  next();
};

// Rate limiting middleware
export const apiLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 100, // Limit each IP to 100 requests per windowMs
  message: { error: 'Too many requests, please try again later.' },
  standardHeaders: true,
  legacyHeaders: false,
});

// CSRF protection middleware
export const csrfProtection = (req: Request, res: Response, next: NextFunction) => {
  // Skip CSRF check for GET, HEAD, OPTIONS, and TRACE methods
  if (['GET', 'HEAD', 'OPTIONS', 'TRACE'].includes(req.method)) {
    return next();
  }

  const csrfToken = req.headers['x-csrf-token'] || req.body._csrf;
  
  if (!csrfToken || !req.session.csrfToken || csrfToken !== req.session.csrfToken) {
    return res.status(403).json({ 
      error: 'Invalid or missing CSRF token' 
    });
  }
  
  next();
};

// Input validation middleware
export const validateRequest = (req: Request, res: Response, next: NextFunction) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ 
      errors: errors.array(),
      message: 'Validation failed' 
    });
  }
  next();
};

// Request ID middleware
export const requestId = (req: Request, res: Response, next: NextFunction) => {
  req.id = req.get('X-Request-Id') || uuidv4();
  next();
};

// Security middleware stack
export const securityMiddleware = [
  // Helmet for basic security headers
  helmet({
    contentSecurityPolicy: false, // We're setting CSP manually
    crossOriginEmbedderPolicy: { policy: 'require-corp' },
    crossOriginOpenerPolicy: { policy: 'same-origin' },
    crossOriginResourcePolicy: { policy: 'same-site' },
    dnsPrefetchControl: { allow: false },
    frameguard: { action: 'deny' },
    hidePoweredBy: true,
    hsts: {
      maxAge: 31536000,
      includeSubDomains: true,
      preload: true,
    },
    ieNoOpen: true,
    noSniff: true,
    referrerPolicy: { policy: 'strict-origin-when-cross-origin' },
    xssFilter: true,
  }),
  
  // Custom security headers
  securityHeaders,
  
  // Request ID
  requestId,
  
  // Body parser
  (req: Request, res: Response, next: NextFunction) => {
    if (req.is('application/json')) {
      express.json({
        limit: '10kb',
        strict: true,
      })(req, res, next);
    } else {
      next();
    }
  },
  
  // URL-encoded parser
  express.urlencoded({
    extended: true,
    limit: '10kb',
    parameterLimit: 10,
  }),
  
  // CSRF protection for non-API routes
  (req: Request, res: Response, next: NextFunction) => {
    if (!req.path.startsWith('/api/')) {
      return csrfProtection(req, res, next);
    }
    next();
  },
];

// Rate limiting for auth endpoints
export const authLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 10, // Limit each IP to 10 requests per windowMs
  message: { error: 'Too many login attempts, please try again later.' },
  standardHeaders: true,
  legacyHeaders: false,
});

// Security headers for API responses
export const apiSecurityHeaders = (req: Request, res: Response, next: NextFunction) => {
  res.set('X-Content-Type-Options', 'nosniff');
  res.set('X-Frame-Options', 'DENY');
  res.set('X-XSS-Protection', '1; mode=block');
  res.set('Strict-Transport-Security', 'max-age=31536000; includeSubDomains');
  next();
};

// Request validation helper
export const validate = (validations: any[]) => {
  return async (req: Request, res: Response, next: NextFunction) => {
    await Promise.all(validations.map(validation => validation.run(req)));

    const errors = validationResult(req);
    if (errors.isEmpty()) {
      return next();
    }

    res.status(400).json({
      errors: errors.array(),
      message: 'Validation failed',
    });
  };
};

// Secure cookie settings
export const secureCookieOptions = {
  httpOnly: true,
  secure: process.env.NODE_ENV === 'production',
  sameSite: 'strict' as const,
  maxAge: 24 * 60 * 60 * 1000, // 1 day
  path: '/',
};

// Generate secure session secret
const generateSessionSecret = (): string => {
  return crypto.randomBytes(32).toString('hex');
};

// Export session configuration
export const sessionConfig = {
  secret: process.env.SESSION_SECRET || generateSessionSecret(),
  name: 'sessionId',
  resave: false,
  saveUninitialized: false,
  cookie: {
    ...secureCookieOptions,
    maxAge: 24 * 60 * 60 * 1000, // 1 day
  },
};

// Log security events
interface SecurityEvent {
  type: string;
  ip: string;
  method: string;
  path: string;
  userAgent?: string;
  timestamp: Date;
  details?: any;
}

export const logSecurityEvent = (event: Omit<SecurityEvent, 'timestamp'>) => {
  const securityEvent: SecurityEvent = {
    ...event,
    timestamp: new Date(),
  };
  
  // In production, this would log to a security monitoring system
  console.warn('[SECURITY EVENT]', securityEvent);
  
  // TODO: Implement actual security event logging
  // This could be sent to a SIEM, security monitoring service, or audit log
};

// Middleware to log security-relevant requests
export const securityLogger = (req: Request, res: Response, next: NextFunction) => {
  // Skip logging for health checks and static assets
  if (req.path === '/health' || req.path.startsWith('/static/')) {
    return next();
  }

  // Log potential security-related events
  if (req.headers['x-forwarded-for']?.toString() !== req.socket.remoteAddress) {
    logSecurityEvent({
      type: 'POTENTIAL_IP_SPOOFING',
      ip: req.ip,
      method: req.method,
      path: req.path,
      userAgent: req.get('user-agent'),
      details: {
        xForwardedFor: req.headers['x-forwarded-for'],
        remoteAddress: req.socket.remoteAddress,
      },
    });
  }

  // Log potential CSRF token issues
  if (req.method !== 'GET' && !req.path.startsWith('/api/')) {
    const csrfToken = req.headers['x-csrf-token'] || req.body?._csrf;
    if (!csrfToken) {
      logSecurityEvent({
        type: 'MISSING_CSRF_TOKEN',
        ip: req.ip,
        method: req.method,
        path: req.path,
        userAgent: req.get('user-agent'),
      });
    }
  }

  next();
};
